cs3api4lab/cs3_file_api.py

```diff
@@ -0,0 +1,278 @@
+	tokens = {}  # map userid [string] to {authentication token, token expiration time}
+
+	def __init__(self, config, log):
+		"""
+		Init module-level variables
+		"""
+		self.ctx['log'] = log
+		self.ctx['chunksize'] = int(config['chunksize'])
+		self.ctx['authtokenvalidity'] = int(config['authtokenvalidity'])
+
+		reva_host = config['revahost']
+
+		# prepare the gRPC connection
+		ch = grpc.insecure_channel(reva_host)
+		self.ctx['cs3stub'] = cs3gw_grpc.GatewayAPIStub(ch)
+		return
+
+	def __authenticate(self, userid):
+
+		"""
+		Obtain a token from Reva for the given userid
+		"""
+
+		# ToDo: use real authentication data
+		# authReq = cs3gw.AuthenticateRequest(type='bearer', client_secret=userid)
+		authReq = cs3gw.AuthenticateRequest(type='basic', client_id='einstein', client_secret='relativity')
+
+		if userid not in self.tokens or self.tokens[userid]['exp'] < time.time():
+			authRes = self.ctx['cs3stub'].Authenticate(authReq)
+			self.tokens[userid] = {'tok': authRes.token, 'exp': time.time() + self.ctx['authtokenvalidity']}
+
+		# piggy back on the opportunity to expire old tokens, but asynchronously
+		# as to not impact the current session: let's use python3.7's coroutines support
+		# asyncio.run(_async_cleanuptokens())
+
+		return self.tokens[userid]['tok']
+
+	def __cs3_reference(self, endpoint, fileid):
+		if fileid[0] == '/':
+			# assume this is a filepath
+			ref = cs3spr.Reference(path=fileid)
+		else:
+			# assume we have an opaque fileid
+			ref = cs3spr.Reference(id=cs3spr.ResourceId(storage_id=endpoint, opaque_id=fileid))
+		return ref
+
+	def stat(self, endpoint, fileid, userid):
+
+		"""
+		Stat a file and returns (size, mtime) as well as other extended info using the given userid as access token.
+		Note that endpoint here means the storage id. Note that fileid can be either a path (which MUST begin with /), or an id (which MUST NOT
+		start with a /).
+		"""
+
+		if endpoint == 'default':
+			raise IOError('A CS3API-compatible storage endpoint must be identified by a storage UUID')
+
+		tstart = time.time()
+
+		ref = self.__cs3_reference(endpoint, fileid)
+
+		statInfo = self.ctx['cs3stub'].Stat(request=cs3sp.StatRequest(ref=ref),
+											metadata=[('x-access-token', self.__authenticate(userid))])
+
+		tend = time.time()
+		self.ctx['log'].info('msg="Invoked stat" fileid="%s" elapsedTimems="%.1f"' % (fileid, (tend - tstart) * 1000))
+
+		if statInfo.status.code == cs3code.CODE_OK:
+			self.ctx['log'].debug('msg="Stat result" data="%s"' % statInfo)
+			return {
+				'inode': statInfo.info.id.storage_id + ':' + statInfo.info.id.opaque_id,
+				'filepath': statInfo.info.path,
+				'userid': statInfo.info.owner.opaque_id,
+				'size': statInfo.info.size,
+				'mtime': statInfo.info.mtime.seconds
+			}
+
+		self.ctx['log'].info('msg="Failed stat" fileid="%s" reason="%s"' % (fileid, statInfo.status.message))
+		raise IOError(statInfo.status.message)
+
+	def stat_x(self, endpoint, fileid, userid):
```

Comment:
Now I would not try to be compatible with WOPI, because Jupyter and WOPI have different requirements and methods of operation. Probability we can remove these methods from API and add if it turns out that Jupyter requires them.